File: 40. Combination Sum II/40. Combination Sum II.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        result = []
        self.backtrack(candidates, target, 0, [], result)
        return result
    
    def backtrack(self, candidates, target, index, current, result):
        if target == 0:
            result.append(list(current))
            return
        
        for i in range(index, len(candidates)):
            if i > index and candidates[i] == candidates[i-1]:
                continue
            if candidates[i] > target:
                break
            current.append(candidates[i])
            self.backtrack(candidates, target - candidates[i], i + 1, current, result)
            current.pop()

File: 40. Combination Sum II/40. Combination Sum II.dart
class Solution {
  List<List<int>> combinationSum2(List<int> candidates, int target) {
    candidates.sort();
    List<List<int>> result = [];
    backtrack(candidates, target, 0, [], result);
    return result;
  }
  
  void backtrack(List<int> candidates, int target, int index, List<int> current, List<List<int>> result) {
    if (target == 0) {
      result.add(List.from(current));
      return;
    }
    
    for (int i = index; i < candidates.length; i++) {
      if (i > index && candidates[i] == candidates[i-1]) continue;
      if (candidates[i] > target) break;
      current.add(candidates[i]);
      backtrack(candidates, target - candidates[i], i + 1, current, result);
      current.removeLast();
    }
  }
}